sw/device/lib/base/hardened.h

```diff
@@ -32,4 +33,41 @@ typedef enum hardened_bool {
   kHardenedBoolFalse = 0x1d4u,
 } hardened_bool_t;
 
+/**
+ * Hardened select of word without branches.
+ *
+ * @param test the value to test
+ * @param a the value returned if 'test' is not 0
+ * @param b the value returned if 'test' is zero
+ *
+ * @return a if test==0 and b otherwise;
+ *
+ * This function is mostly useful when test == 0 is a good result and
+ * hardening is required to prevent easy generation of 'b'.
+ *
+ * Example:
+ *   hardened_select_if_zero(test, kHardenedBoolFalse, kHardenedBoolTrue)
+ *
+ * We need a form where skipping instruction will not allow to
+ * produce kHardenedBoolTrue, on RV32 try to get 'snez' instead of 'seqz'
+ * as snez will transform non-zero reg into non-zero, so skipping it
+ * won't cause an inversion of value.
+ * Constant 7 is choosen to prevent input values in 0xfffffff0 - 0xffffffff
+ * range to be converted to proper constant by skipping snez & neg.
+ */
+static inline uintptr_t hardened_select_if_zero(uintptr_t test, uintptr_t a,
```

Comment:
I may have mentioned this before, but `static inline` is banned. Instead, you should place `extern` declarations in `hardened.c` (and create it if it's missing). This is an artifact of C getting inline wrong.

Reply:
i can't use inline - it fails to compile C++ tests. i get hardened_select_if_zero() is undefined at link stage for host test.

Reply:
This was resolved offline. C inline is kind of a disaster but hopefully you have a bit more context now.